src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,406 @@
+    {
+    }
+
+    /**
+     * From the standard: rebind is only optional (provided by std::allocator_traits) if this
+     * allocator is a template of the form SomeAllocator<T, Args>, where Args is zero or more
+     * additional template type parameters.
+     *
+     * Since we use a size_t as additional argument, it's *not* optional.
+     * @see https://en.cppreference.com/w/cpp/named_req/Allocator#cite_note-2
+     */
+    template <typename U>
+    struct rebind {
+        using other = Allocator<U, ALLOCATION_SIZE_BYTES>;
+    };
+
+    /**
+     * Allocates n entries of the given type.
+     */
+    T* allocate(size_t n)
+    {
+        // Forward all allocations to the memory_resource
+        return m_memory_resource->template Allocate<T>(n);
+    }
+
+    /**
+     * Deallocates n entries of the given type.
+     */
+    void deallocate(T* p, size_t n)
+    {
+        m_memory_resource->template Deallocate<T>(p, n);
+    }
+
+private:
+    //! Stateful allocator, where the state is a simple pointer that can be cheaply copied.
+    MemoryResource<ALLOCATION_SIZE_BYTES>* m_memory_resource;
+};
+
+
+/**
+ * Since Allocator is stateful, comparison with another one only returns true if it uses the same
+ * memory_resource.
+ */
+template <typename T, typename U, size_t CS>
+bool operator==(const Allocator<T, CS>& a, const Allocator<U, CS>& b) noexcept
+{
+    // "Equality of an allocator is determined through the ability of allocating memory with one
+    // allocator and deallocating it with another." - Jonathan Müller
+    // See https://www.foonathan.net/2015/10/allocatorawarecontainer-propagation-pitfalls/
+    //
+    // For us that is the case when both allocators use the same memory resource.
+    return a.m_memory_resource == b.m_memory_resource;
+}
+
+template <typename T, typename U, size_t CS>
+bool operator!=(const Allocator<T, CS>& a, const Allocator<U, CS>& b) noexcept
+{
+    return !(a == b);
+}
+
+/**
+ * Helper to create std::unordered_map which uses the node_allocator.
+ *
+ * This calculates the size of the container's internally used node correctly for all supported
+ * platforms, which is also asserted by the unit tests.
+ */
+template <typename Key, typename Value, typename Hash = std::hash<Key>, typename Equals = std::equal_to<Key>>
+class UnorderedMapFactory
+{
+public:
+    static constexpr size_t ALLOCATION_SIZE_BYTES = memusage::NodeSize<std::unordered_map<Key, Value, Hash, Equals>>::Value();
+    using MemoryResourceType = MemoryResource<ALLOCATION_SIZE_BYTES>;
+    using AllocatorType = Allocator<std::pair<const Key, Value>, ALLOCATION_SIZE_BYTES>;
+    using ContainerType = std::unordered_map<Key, Value, Hash, Equals, AllocatorType>;
+
+    /**
+     * Creates the std::unordered_map container, and asserts that the specified memory_resource is
+     * correct.
+     */
+    [[nodiscard]] static ContainerType CreateContainer(MemoryResourceType* memory_resource)
```

Comment:
I'd like to keep the `CreateContainer` here because otherwise constructing an unordered_map is quite a lot of non-obvious code, like first argument is always 0, then create a temporary hash of the correct typ, equals, and finally the Allocator.